lib/Scheduler/Scheduler.cpp

```diff
@@ -0,0 +1,116 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Scheduler/Scheduler.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "ttmlir/Dialect/TTNN/IR/TTNNOpsDialect.h.inc"
+#include <mlir/IR/BuiltinOps.h>
+#include <mlir/IR/Operation.h>
+#include <mlir/Interfaces/DestinationStyleOpInterface.h>
+
+namespace mlir::tt::scheduler {
+
+bool isDpsOperand(mlir::Operation *op, const mlir::OpOperand *operand) {
+  // Check if the operation implements DestinationStyleOpInterface
+  if (auto dpsOp = dyn_cast<mlir::DestinationStyleOpInterface>(op)) {
+    if (dpsOp.getDpsInitOperand(0) == operand) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool isTTNNOp(mlir::Operation *op) {
+  return isa<ttnn::TTNNDialect>(op->getDialect());
+}
+
+bool isDpsOp(mlir::Operation *op) {
+  return isa<mlir::DestinationStyleOpInterface>(op);
+}
+
+// Constructor taking an MLIR Operation (or a module)
+// and generates the dependencies between the operations
+Scheduler::Scheduler(mlir::Operation *root) {
+  // Check that the root operation is a module
+  assert(isa<mlir::ModuleOp>(root) && "Root operation must be a module");
+  mlir::ModuleOp module = dyn_cast<mlir::ModuleOp>(root);
+
+  // Check that the module has a single function
+  auto ops = module.getOps<func::FuncOp>();
+  assert(std::distance(ops.begin(), ops.end()) == 1 &&
+         "Module must have a single function");
+
+  func::FuncOp func = *ops.begin();
+
+  // Init the dependencies map of all TTNN ops
+  // which implement DestinationStyleOpInterface
+  // We check for DPS because all operations which do
+  // some computation in TTNN implement this interface
+  for (auto &op : func.getOps()) {
+    if (isTTNNOp(&op) && isDpsOp(&op)) {
+      dependencies[&op] = {};
+      unscheduledOps.insert(&op);
+    }
+  }
+
+  for (auto &op : func.getOps()) {
+    // Skip non TTNN operations
+    // Skip operations which do not implement DestinationStyleOpInterface
+    if (!isTTNNOp(&op) || !isDpsOp(&op)) {
```

Comment:
My understanding is that scheduler should work on TTIR. @nobradovictt Is that correct?

Reply:
Yes it needs to work on TTIR so it can be used within Optimizer before lowering to TTNN.